File: src/vrpn_gravity_projection/src/twist_echo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
订阅 /vrpn_client_node/Rigid/twist 话题
对数据进行变换：第三个取负，然后交换第二和第三个位置
即: (x, y, z) -> (x, -z, y)
"""

import rospy
from geometry_msgs.msg import TwistStamped

def twist_callback(msg):
    # 原始数据
    lx = msg.twist.linear.x
    ly = msg.twist.linear.y
    lz = msg.twist.linear.z
    
    ax = msg.twist.angular.x
    ay = msg.twist.angular.y
    az = msg.twist.angular.z
    
    # 变换: 第三个取负，然后交换第二和第三
    # (x, y, z) -> (x, -z, y)
    new_lx = lx
    new_ly = -lz
    new_lz = ly
    
    new_ax = ax
    new_ay = -az
    new_az = ay
    
    print("---")
    print("linear:")
    print("  x: %.6f" % new_lx)
    print("  y: %.6f" % new_ly)
    print("  z: %.6f" % new_lz)
    print("angular:")
    print("  x: %.6f" % new_ax)
    print("  y: %.6f" % new_ay)
    print("  z: %.6f" % new_az)

if __name__ == '__main__':
    rospy.init_node('twist_echo_transformed', anonymous=True)
    rospy.Subscriber('/vrpn_client_node/Rigid/twist', TwistStamped, twist_callback)
    print("Listening to /vrpn_client_node/Rigid/twist ...")
    print("Transform: (x, y, z) -> (x, -z, y)")
    rospy.spin()

File: src/vrpn_gravity_projection/src/twist_echo.cpp
#include <ros/ros.h>
#include <geometry_msgs/TwistStamped.h>

void twistCallback(const geometry_msgs::TwistStamped::ConstPtr& msg)
{
    // 原始数据
    double lx = msg->twist.linear.x;
    double ly = msg->twist.linear.y;
    double lz = msg->twist.linear.z;
    
    double ax = msg->twist.angular.x;
    double ay = msg->twist.angular.y;
    double az = msg->twist.angular.z;
    
    // 变换: 第三个取负，然后交换第二和第三
    // (x, y, z) -> (x, -z, y)
    double new_lx = lx;
    double new_ly = -lz;
    double new_lz = ly;
    
    double new_ax = ax;
    double new_ay = -az;
    double new_az = ay;
    
    printf("---\n");
    printf("linear:\n");
    printf("  x: %f\n", new_lx);
    printf("  y: %f\n", new_ly);
    printf("  z: %f\n", new_lz);
    printf("angular:\n");
    printf("  x: %f\n", new_ax);
    printf("  y: %f\n", new_ay);
    printf("  z: %f\n", new_az);
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "twist_echo_transformed");
    ros::NodeHandle nh;
    
    printf("Listening to /vrpn_client_node/Rigid/twist ...\n");
    printf("Transform: (x, y, z) -> (x, -z, y)\n");
    
    ros::Subscriber sub = nh.subscribe("/vrpn_client_node/Rigid/twist", 10, twistCallback);
    
    ros::spin();
    return 0;
}
